src/transform/src/attribute/cardinality.rs

```diff
@@ -0,0 +1,646 @@
+    /// The total cardinality of a given global id
+    Id(GlobalId),
+    /// The inverse of the number of distinct keys in the index held in the given column, i.e., 1/|# of distinct keys|
+    ///
+    /// TODO(mgree): need to correlate this back to a given global table, to feed in statistics (or have a sub-attribute for collecting this)
+    Index(usize),
+    /// An unbound local or other unknown quantity
+    Unknown,
+}
+
+/// SymbolicExpressions specialized to factorizer variables
+pub type SymExp = SymbolicExpression<FactorizerVariable>;
+
+/// A `Factorizer` computes selectivity factors
+pub trait Factorizer {
+    /// Compute selectivity for the flat map of `tf`
+    fn flat_map(&self, tf: &TableFunc, input: &SymExp) -> SymExp;
+    /// Computes selectivity of the predicate `expr`, given that `indexed_columns` are indexed
+    ///
+    /// The result should be in the range (0, 1.0]
+    fn predicate(&self, expr: &MirScalarExpr, indexed_columns: &BTreeSet<usize>) -> SymExp;
+    /// Computes selectivity for a filter
+    fn filter(
+        &self,
+        predicates: &Vec<MirScalarExpr>,
+        keys: &Vec<Vec<usize>>,
+        input: &SymExp,
+    ) -> SymExp;
+    /// Computes selectivity for a join; the cardinality estimate for each input is paired with the keys on that input
+    ///
+    /// `indexed_columns` maps column references (with indices) to their index in `inputs`
+    fn join(
+        &self,
+        equivalences: &Vec<Vec<MirScalarExpr>>,
+        implementation: &JoinImplementation,
+        indexed_columns: BTreeMap<usize, usize>,
+        inputs: Vec<&SymExp>,
+    ) -> SymExp;
+    /// Computes selectivity for a reduce
+    fn reduce(
+        &self,
+        group_key: &Vec<MirScalarExpr>,
+        expected_group_size: &Option<u64>,
+        input: &SymExp,
+    ) -> SymExp;
+    /// Computes selectivity for a topk
+    fn topk(
+        &self,
+        group_key: &Vec<usize>,
+        limit: &Option<usize>,
+        expected_group_size: &Option<u64>,
+        input: &SymExp,
+    ) -> SymExp;
+    /// Computes slectivity for a threshold
+    fn threshold(&self, input: &SymExp) -> SymExp;
+}
+
+/// The simplest possible `Factorizer` that aims to generate worst-case, upper-bound cardinalities
+#[derive(Debug)]
+pub struct WorstCaseFactorizer {
+    /// cardinalities for each `GlobalId` and its unique values
+    pub cardinalities: BTreeMap<FactorizerVariable, usize>,
+}
+
+/// The default selectivity for predicates we know nothing about.
+///
+/// It is safe to use this instead of `FactorizerVariable::Index(col)`.
+pub const WORST_CASE_SELECTIVITY: f64 = 0.1;
+
+impl Factorizer for WorstCaseFactorizer {
+    fn flat_map(&self, tf: &TableFunc, input: &SymExp) -> SymExp {
+        match tf {
+            TableFunc::GenerateSeriesInt32
+            | TableFunc::GenerateSeriesInt64
+            | TableFunc::GenerateSeriesTimestamp
+            | TableFunc::GenerateSeriesTimestampTz => input.clone(),
```

Comment:
What's the reason to consider `GenerateSeries~` to have explosion factor `1.0`? Intuitively I would think that this should have the be the case as the default at the end of this match statement.

Reply:
Almost surely because I misunderstood how these operators work.